routes/api/v1/boring_query.js

```diff
@@ -0,0 +1,133 @@
+        const appKey = process.env.EDAMAM_KEY
+        url = `https://api.edamam.com/search?q=${searchQuery}&app_id=${appId}&app_key=${appKey}&calories=2000-999999&to=30`
+
+        //fetch new recipes from edamam api
+        getRecipes(url)
+        .then(recipeResponse => {
+          //create BoringQuery, Recipes, and BoringQueryRecipes in database
+          return saveBoringRecipes(recipeResponse, searchQuery);
+        })
+        .then(recipes => {
+          //Send newly retrieved recipes. Sort by highest calorie total and only top 10 results.
+          recipes.sort((a, b) => b.calories - a.calories);
+          recipes = recipes.slice(0, 10);
+          res.status(200).send(JSON.stringify(recipes))
+        })
+        .catch(error => {
+          res.status(404).send({error: error})
+        })
+      //If an existing query is found
+      } else {
+        // find the recipes that are already associated with the existing query
+        Recipe.findAll({
+          include: [{
+            model: BoringQueryRecipe,
+            attributes: [],
+            where: {
+              BoringQueryId: query.id
+            }
+          }],
+          order: [['calories', 'DESC']],
+          limit: 10
+        })
+        .then(recipes => {
+          //Send retrieved recipes.
+          res.status(200).send(JSON.stringify(recipes))
+        })
+        .catch(error => {
+          res.status(404).send({error: error})
+        })
+      }
+    })
+    .catch(error => {
+      res.status(404).send({error: "Bad search query."})
+    })
+  }
+})
+
+function getRecipes(url){
+  return new Promise((resolve, reject) => {
+    fetch(url)
+    .then(response => {
+      if (response.status = 200){
+        resolve(response.json())
+      }else{
+        reject(result.error_message)
+      }
+    })
+  })
+};
+
+function saveBoringRecipes(recipeResponse, searchQuery){
+  return new Promise((resolve, reject) => {
+    BoringQuery.create({
+      query: searchQuery
+    })
+    .then(query => {
+      resolve(parseRecipes(recipeResponse, query));
+    })
+    .catch(error => {
+      reject(error)
+    })
+  })
+};
+
+function parseRecipes(recipeResponse, query) {
+  return Promise.all(recipeResponse.hits.map(recipe => {
+    return saveRecipe(recipe, query);
+  }))
+};
+
+function saveRecipe(recipe, query){
+  return new Promise((resolve, reject) => {
+    Recipe.findOrCreate({
+      where: {name: recipe.recipe.label},
+      defaults: {
+        url: recipe.recipe.url,
+        yield: recipe.recipe.yield,
+        calories: Math.round((recipe.recipe.calories / recipe.recipe.yield)),
+        image: recipe.recipe.image,
+        totalTime: recipe.recipe.totalTime
+      }
+    })
+    .then(recipe => {
+      BoringQueryRecipe.create({
+        BoringQueryId: query.id,
+        RecipeId: recipe[0].id
+      })
+      .then(() => {
+        resolve(recipe[0])
+      })
+    })
+    .catch((error) => {
+      reject(error)
```

Comment:
Does reject kick it out of this loop and into the bigger loop where the catch sends a 500 error?

Reply:
Basically 'resolve' is always saying 'this passes/operates as expected and here's the variable to pass, while reject is always saying this did not go as expected and heres the error. Resolve will force the original function to follow the ".then" path, while reject will just force it to follow the ".catch" path. It doesn't imply a 500 error. In fact, nothing in our code will ever (hardly ever?) result in a 500. Thats like an out of our control network error. We should usually be sending 400 errors.

Reply:
But mostly yes, it kicks out into the catch.